libyul/optimiser/UnusedFunctionParameterPruner.cpp

```diff
@@ -0,0 +1,136 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+/**
+ * UnusedFunctionParameterPruner: Optimiser step that removes unused parameters from function
+ * definition.
+ */
+
+#include <libyul/optimiser/UnusedFunctionParameterPruner.h>
+#include <libyul/optimiser/UnusedFunctionsCommon.h>
+#include <libyul/optimiser/OptimiserStep.h>
+#include <libyul/optimiser/NameCollector.h>
+#include <libyul/optimiser/NameDisplacer.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/YulString.h>
+#include <libyul/AsmData.h>
+#include <libyul/Dialect.h>
+
+#include <libsolutil/CommonData.h>
+
+#include <algorithm>
+#include <optional>
+#include <variant>
+
+using namespace std;
+using namespace solidity::util;
+using namespace solidity::yul;
+
+namespace
+{
+
+bool anyFalse(vector<bool> const& _mask)
+{
+	return any_of(_mask.begin(), _mask.end(), [](bool b){ return !b; });
+}
+
+template <typename K, typename V>
+optional<V> optionalFind(map<K, V> const& _map, K const& _key)
+{
+	auto it = _map.find(_key);
+
+	if (it != _map.end())
+		return it->second;
+
+	return nullopt;
+}
+
+
+} // anonymous namespace
+
+void UnusedFunctionParameterPruner::run(OptimiserStepContext& _context, Block& _block)
+{
+
+	map<YulString, size_t> references = ReferencesCounter::countReferences(_block);
+	auto used = [&](auto v) -> bool { return references.count(v.name); };
+
+	// Function name and a boolean mask, where `false` at index `i` indicates that the function
+	// argument at index `i` in `FunctionDefinition::parameters` is unused inside the function body.
+	map<YulString, vector<bool>> usedParameters;
```

Comment:
If you change this to a `map<YulString, vector<bool>[2]>` and include both the parameters and the return variables - would this simplify the code? I would say that you won't need `optionalFind` below and the if statements in https://github.com/ethereum/solidity/pull/9397/files#diff-b67936fa269149fe7e133f276f7cab60R107 can be removed (i.e. only the "true" case will be left).

Reply:
I changed it to a pair of vectors.